soroban-env-common/src/array.rs

```diff
@@ -1,24 +1,20 @@
 use crate::xdr::ScHostValErrorCode;
+use core::borrow::Borrow;
 use stellar_xdr::ScObjectType;
 
-use crate::{
-    ConversionError, Env, Object, RawVal, RawValConvertible, Status, TryFromVal, TryIntoVal,
-};
-
-// TODO: these conversions happen as RawVal, but they actually take and produce
-// Objects; consider making the signatures tighter.
+use crate::{ConversionError, Env, Object, RawVal, RawValConvertible, Status, TryFromVal};
 
 impl<E: Env, const N: usize> TryFromVal<E, RawVal> for [u8; N] {
     type Error = Status;
 
-    fn try_from_val(env: &E, val: RawVal) -> Result<Self, Self::Error> {
+    fn try_from_val(env: &E, val: impl Borrow<RawVal>) -> Result<Self, Self::Error> {
+        let val = *val.borrow();
```

Comment:
I think using Borrow<> is a great idea. One thing I am curious about is if this will result in code size changes. I saw using refs for some conversions increase WASM binary size, although not significantly. I don't think we should worry about this much, just wanted to share it.

Reply:
Looking from this change set. It appears the usages of `Borrow<T>` are exactly like a `&T`, i.e. there is no additional `impl Borrow<T>` for any type (other than the implicit `T`, `&T`, `Box<T>` etc)? So I guess from the binary size perspective there should not be a difference.

I wonder then, what problem doe `impl Borrow<T>` (instead of `AsRef<T>`) solve?

Reply:
`Borrow` and `AsRef` do similar things but are subtly different. `Borrow` is implemented on all types automatically. 

Part of the contract of Borrow is that `Eq`, `Ord`, and `Hash` have to work identically for `T` and for the borrowed value of T.

AsRef does not have the same requirements. Any type can offer an AsRef of another type, even if those two types are very very different.

There's some details about it here: https://doc.rust-lang.org/std/borrow/trait.Borrow.html

Reply:
Right, I get those differences. 
I don't think (at least from looking at this change set) that we are replying on any other logic than those automatic implementations. So `*val.borrow()` is just `&val`, which should lead to no difference in compiled binary code (and also wasm code size)?

Reply:
My question was, if the `impl Borrow` are just automatic implementations, and we are only passing in `T` and `&T` as arguments in most cases, what problems did `impl Borrow` solve, there must be corner cases?

Reply:
The use of `impl Borrow<T>` here is specifically (and solely) to allow us to write a single impl of TryFromVal that will work for _either_ T or &T arguments. We previously had -- in a variety of places, some impls but not all -- duplicate impls for both of those cases. The whole point of this PR is "reduce the number of impls".

Reply:
I see, thanks for the clarification.

Reply:
@jayz22 I actually stared at this answer for a few minutes and wondered again how much pain it'd be to go with just &T everywhere, and I made a change to do that and it's .. not too bad. Actually lets me remove some of the additional type-inference helpers I put in before. It's very slightly less ergonomic when using `T::try_from_val` since you have to `&arg` the arguments, but for the more common case of calling `x.try_into_val()` it makes no difference since I already have the signature of `try_into_val(&self, e)` having that `&self` parameter -- rust will auto-ref the self arg.

Considering this has even less cognitive load and seems better behaved in terms of type inference I'm going to try going along with it. I think the ergonomic hit will not be so bad. @leighmcculloch mentioned there might be a slight codesize hit which I'll .. try to quantify, but again at this point I'm primarily concerned with getting this code under _control_ so it makes sense, is smaller, more consistent and maintainable.